feat issue-#23: 마이페이지 관련 기능 구현 및 토큰 식별 email로 변경

BOK-MINJEONG opened the pull request:
## #️⃣ 관련 이슈
close #28 
close #23 

## 📝작업 내용
(피그마 페이지명 기준입니다)
- 마이페이지
- 마이페이지_내정보
- 마이페이지_내정보_이름 변경
- 마이페이지_프로필 사진 변경
- 회원가입 시 기본 프로필로 설정되게 했습니다. 이미지는 버킷에 올렸습니다!
<br>   

- login, logout, generateNewAccessToken, deleteUser에서 username으로 식별하던 부분 email로 식별하게 변경했습니다
- jwt 토큰 발급, 유저 식별을 email 기준으로 작동하게 수정했습니다   
~~`Parameter 0 of constructor in com.canvi.hama.common.util.RedisUtil required a single bean, but 2 were found:` 에러가 발생해서 RedisConfig에 @Primary 어노테이션 추가했습니다~~
- RedisConfig 삭제했습니다!

---
> Swagger 설정에 인증 헤더를 추가하여 JWT 토큰을 포함할 수 있도록 SwaggerConfig 수정했습니다.
> 로그인 후 발급받은 accessToken을 Authorize -> value 탭에 넣어 사용하시면 됩니다!
![image](https://github.com/user-attachments/assets/0436c766-0ae4-4ef0-a211-432ee02ba1e1)

BOK-MINJEONG committed:
feat issue-#23: 마이페이지 관련 기능 구현 및 로그인 요청 email로 변경

- 마이페이지
- 마이페이지_내정보
- 마이페이지_내정보_이름 변경
- 마이페이지_프로필 사진 변경

- jwt 토큰 발급, 유저 식별 username -> email로 변경

zhy2on commented on src/main/java/com/canvi/hama/domain/diary/entity/Image.java:
```diff
@@ -17,7 +17,6 @@ public class Image extends BaseEntity {
     @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
 
-    @JsonBackReference
```
혹시 여기는 왜 제거하신 걸까요?? 제가 `@JsonBackReference`의 사용법을 잘 몰라서요..!

BOK-MINJEONG replied:
diary와 image가 서로 참조하면서 순환 참조 문제가 발생해서 사용했습니다.! 서로의 정보를 계속 참조하면 `StackOverFlowError`가 발생하거든요.!

근데 이런 직렬화 순환 참조 문제는 '양방향 매핑' + '**entity 직접 반환**' 때문에 생긴다고 해요
dto 객체를 만들어 반환하면 문제를 방지할 수 있어서 어노테이션 제거했습니다!

항상 이 문제 생길 때마다 헷갈려서 저도 찾아보면서 하게 되네요..!
[https://blogshine.tistory.com/436](url)

zhy2on commented on src/main/java/com/canvi/hama/domain/diary/service/DiaryService.java:
```diff
@@ -22,29 +24,23 @@
-        Optional<Diary> diaryOptional = diaryRepository.findByUserIdAndDiaryDate(diaryRequest.getUserId(), diaryRequest.getDiaryDate());
-        if (diaryOptional.isPresent()) {
+        User user = userService.getUserByUserId(diaryRequest.getUserId());
+        boolean isExistDiary = diaryRepository.existsByUserIdAndDiaryDate(diaryRequest.getUserId(), diaryRequest.getDiaryDate());
```
👍👍

zhy2on commented on src/main/java/com/canvi/hama/domain/user/entity/User.java:
```diff
@@ -37,21 +35,27 @@ public class User extends BaseEntity {
 
     @Builder(access = AccessLevel.PRIVATE)
-    private User(String username, String password, String email) {
+    private User(String username, String password, String email, String profile) {
```
`DEFAULT_PROFILE="https://canvi-hama-bucket.s3.ap-northeast-2.amazonaws.com/hama_profile.png"` 이렇게 매직넘버 사용해서 엔티티에서 초기화 해도 좋을 것 같습니다!

zhy2on commented on src/main/java/com/canvi/hama/common/config/RedisConfig.java:
```diff
@@ -23,6 +24,7 @@ public RedisConnectionFactory redisConnectionFactory() {
         return new LettuceConnectionFactory(host, port);
     }
 
+    @Primary
```
오 처음 알게 된 건데 빈을 우선적으로 적용하는 거군요

근데 애초에 RedisTemplate 빈이 왜 중복되지? 하고 찾아보니까 spring boot에서 자동으로 생성해주는 거였네요...
RedisConfig를 삭제해도 application.yml 설정대로 잘 작동을 하네요..🥲 파일 자체를 삭제하면 될 것 같습니다!

https://developer-nyong.tistory.com/21
https://jronin.tistory.com/126

<img width="732" alt="image" src="https://github.com/user-attachments/assets/f8cfb08b-6cb4-43dd-9d36-3f0955e83fec">

BOK-MINJEONG replied:
spring boot에서 자동으로 생성해서 중복된다는 에러가 생긴 거였군요..

저는 깊이 생각해보지 않고 로그에 찍힌 내용만 적용했네요ㅠ..
좋은 블로그 추천 감사합니다!👍

zhy2on commented on src/main/java/com/canvi/hama/domain/user/dto/UpdateNameRequest.java:
```diff
@@ -0,0 +1,9 @@
+
+import lombok.Getter;
+
+@Getter
```
```java
@Getter
public class UpdateNameRequest (
    @NotBlank(message = "유저 아이디가 비어있습니다.")
    Long userId,
    
    @NotBlank(message = "변경할 이름이 비어있습니다.")
    String username
) {
}
```

dto는 record 사용으로 바꾸면 좋을 것 같습니다!
컨트롤러에서 사용할 때 `@Valid`를 이용해서 유효성 검사를 하면 좋을 것 같아요!

zhy2on commented on src/main/java/com/canvi/hama/domain/user/dto/MyPageInfoResponse.java:
```diff
@@ -0,0 +1,17 @@
+
+@Getter
+@NoArgsConstructor
+public class MyPageInfoResponse {
```
여기도 record 사용으로 변경하면 좋을 것 같습니다!

zhy2on commented on src/main/java/com/canvi/hama/domain/user/controller/UserController.java:
혹시 명시적으로 나타내기 위해서 따로 try-catch문을 사용하신 걸까요??
그게 아니라면 글로벌 익셉션 핸들러에서 BaseException을 처리 하게 해놔서 따로 try catch를 안 해도 될 것 같습니다!

BOK-MINJEONG replied:
따로 의도를 가지고 사용한 건 아니었습니다! 수정하겠습니다!

zhy2on commented on src/main/java/com/canvi/hama/domain/user/controller/UserController.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @NameUpdateApi
+    @PatchMapping("/name")
```
오... `patch`를 처음 알았네요
resource의 전체를 교체하는 게 `put`이고
부분만 교체하는 데 쓰이는 게 `patch` 군요!

새로운 메서드 알아갑니다🙇
https://programmer93.tistory.com/39

zhy2on commented on src/main/java/com/canvi/hama/domain/user/controller/UserController.java:
api 너무 깔끔합니다bb
근데 로그인 이후 이뤄져야 하는 과정이기 때문에 `@AuthenticationPrincipal UserDetails userDetails`를 사용하는 게 좋을 것 같아요!

```java
@GetMapping("/me")
public BaseResponse<MyPageResponse> getMyPage(@AuthenticationPrincipal UserDetails userDetails)
```

zhy2on commented on src/main/java/com/canvi/hama/domain/user/service/UserService.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Transactional
+    public void updateName(UpdateNameRequest updateNameRequest) {
+        User user = getUserByUserId(updateNameRequest.getUserId());
```
``User user = getUserFromUserDetails(userDetails);``
userDetails를 사용한다면 이렇게 바꾸면 좋을 것 같습니다!

zhy2on commented on src/main/java/com/canvi/hama/domain/user/entity/User.java:
```diff
@@ -37,21 +35,27 @@ public class User extends BaseEntity {
     @NotNull
     private int credits = DEFAULT_CREDITS;
 
+    @Setter
```
https://velog.io/@langoustine/setter-지양-이유
https://jeongkyun-it.tistory.com/221

entity에선 setter 사용을 지양하는 게 좋더라고요!
의미 있는 메서드로 대체하는 것이 좋다고 해서
updateProfile 메서드를 만들어서 사용하면 좋을 것 같습니다!

zhy2on commented on src/main/java/com/canvi/hama/domain/user/controller/UserController.java:
```diff
@@ -0,0 +1,68 @@
+@RestController
+@RequestMapping("/api/user")
+@RequiredArgsConstructor
+@Slf4j
```
진짜 사소한 거긴 하지만...

<img width="864" alt="image" src="https://github.com/user-attachments/assets/dd3d27f0-7cb9-4a10-ad7f-729cbdd25b3f">

보통 `@Service` `@RestController` 같이 클래스 역할을 나타내는 어노테이션이 가장 상단에 오는 게 가독성이 좋다고 하더라고요!
사실 크게 신경 쓸 부분은 아니긴 한데 어디서 봤던 거라 참고하실만할까 싶어 남겨봅니다!

zhy2on reviewed (approved):
수고하셨습니다ㅠㅠ!
처음 알아가는 내용도 있어서 좋았고
깔끔하게 작성해주셔서 보기 편했습니다 !!

BOK-MINJEONG committed:
fix: config 수정

- RedisConfig.java 삭제
- Swagger 설정에 인증 헤더를 추가하여 JWT 토큰을 포함할 수 있도록 SwaggerConfig 수정

BOK-MINJEONG committed:
fix: DEFAULT_PROFILE 상수로 선언

BOK-MINJEONG committed:
refactor: dto record로 변경 및 불필요한 구문 삭제

BOK-MINJEONG committed:
fix: record 수정

- 어노테이션 삭제

BOK-MINJEONG committed:
fix: diary, image 롤백

BOK-MINJEONG committed:
fix: 충돌 해결
